### ISSUE TITLE
Blocking apps?

### ISSUE THREAD

**Issue by OffifialMITX** [blokadaorg/blokada#49]
I'd like to completely block internet access for certain apps. Would that work?

**Comment by kar** [comment]
Not as of now, but it's a good feature request.

**Comment by jeroen7s** [comment]
:+1: for this request

**Comment by kar** [comment]
> Hi there, amazing app you got here. I was wondering if you could add an option to block a specific app from accessing the internet. Sometimes some apps request access to certain things on my device, that it doesn't need. I am on Android 5.1.1. so I can't deny specific permissions, so having an option on your app to block it from accessing the internet would be great. In case you already have this option, my bad, I'm new to this app. Let me know how to activate it. Thanks :)

**Comment by jerryn70** [comment]
+1👍

**Comment by GambaJo** [comment]
This would be great. But I would like to add a single app to the blacklist or all. All would be very useful, because I want to black all apps and only give some of them a permission to access the internet. Also for the system apps.

**Comment by maverick74** [comment]
I would like the same as @GambaJo !!!

All apps access to internet blocked by default and only allow the ones I really need!!! 

Naturally, the ones allowed should still have their ads blocked!!!

**Comment by rpeter85** [comment]
You are an advanced user. Most of the users aren't. If we would block every internet access, the users would remove the app within 5 sec.
So, what about allowing a few and you can decide what should be blocked?

**Comment by maverick74** [comment]
@rpeter85 I really don't care what the defaults are as long as one can change them in the configs!

In fact it would indeed be better if it would just block ads by default and then, if the user wants, be able to block the apps.

Ideally to block apps I think the user should first decide if it wants to work with a white list (where by default all apps are allowed to access internet except the ones in the list) or with a black list (where by default all apps are blocked from access internet except the ones in the list).

Again, as long as one can easily configure it, defaults as not important for advanced users :)

**Comment by GambaJo** [comment]
@rpeter85 It depends on how you implement this. If it's transparent, what's happening, then I think, it will be ok. Some system apps shouldn't be blocked, but you can show a hint.
I know this feature from NetGuard. You can configure the app, that every app is blocked, even the new installed. And then the user can decide, if the app should have access.

**Comment by maverick74** [comment]
Yes, NetGuard has this and i think it's well implemented (unfortunately some features are pay-for only...)

**Comment by colbycdev** [comment]
This would be a handy feature for those who only or mostly use mobile data. If I can say "only apps x, y, z get internet" or "<insert data hog> app gets no internet" that would be great.
Also, if an app has the internet permissions but doesn't use it in a beneficial way(such as ones that only report your info to a server) you could blacklist it. This way you could also integrate with existing blacklists

**Comment by maverick74** [comment]
I, personally, would like to set that no apps get internet access except the ones I say so.
From the ones that are allowed, I would like those to get their ads blocked (while retaining Internet access to the useful features that really need it)

**Comment by RichyHBM** [comment]
My view would be that by default all apps are under the Blokada filter, then you would have a drop-down that lets you set between allow all (already in the app I believe), filter network (i.e. use Blokada filtering), and Deny all (i.e. block all internet access)

**Comment by Esokrates** [comment]
I want to outline though, that it would be desirable to implement this feature like in Netguard: Being able to whitelist/blacklist internet access for wifi / mobile data connection separately, so I can block apps wasting my data plan while allowing everything on wifi.

Use case: For example most children have tablets/phones today and it would be desirable to only whitelist mobile data for select applications, like messaging apps. So they can safely use apps without wasting money due to them not understanding / forgetting they are not on wifi.

Maybe you could borrow some code from Netguard? The Netguard Dev btw has no intentions to allow automatic filter updates and the like, so partly redoing Netguard is really appreciated due to the dev of Netguard having a different vision.

**Comment by okan35** [comment]
Hello, is anyone working on this as I could definitely check this and try to implement it as I have been using blokada for a long time and want to block some apps internet access,so I don't want to use another app for this.

**Comment by gkrishnaks** [comment]
Netguard already has this feature. Currently blokada is lightweight , low on ram and battery.  I wonder if this feature will increase ram and battery consumption. Maybe a way to have it default ON, but toggle OFF completely if user wants to save her battery or ram.

**Comment by Esokrates** [comment]
> Netguard already has this feature. Currently blokada is lightweight , low on ram and battery. I wonder if this feature will increase ram and battery consumption. Maybe a way to have it default ON, but toggle OFF feature button to switch it off completely if user wants to save her battery or ram. Although I am not sure if that will save memory or battery either. Just thinking about it!

While Netguard has that feature, Netguard is not a superset of blokada in terms of functionality, especially not regarding filter lists / adblocking. In Netguard you need to manage everything about your filters manually (and you can only specify one filter list), while in blokada you can have multiple preconfigured lists that are updated automatically by the app.

**Comment by gkrishnaks** [comment]
Yes, I'm aware of the differences. 
I had been a long time user of netguard app, recently switched to Blokada 5 months ago.

**Comment by colbycdev** [comment]
The battery usage on netguard is pretty limited from my use, so that feature integration should be relatively lightweight
On the black/whitelist choice, I think it should be blacklist by default and switchable to whitelisting

**Comment by ka223** [comment]
The firewall options (block, mobile, wifi, screen on ...) are unique on netguard.
Maybe something similar can be implemented here too?

**Comment by beerisgood** [comment]
LineageOS users can already config that app-based

**Comment by maverick74** [comment]
@okan35 and @kar is there any news in this issue???

Is okan35 working on any front in this?

**Comment by okan35** [comment]
I haven't worked on it because code is kind of hard to understand and the feature is  not easy to implement also.

**Comment by infinitytec** [comment]
It would also be nice to set data limits for apps.

**Comment by iriseecz** [comment]
I tried to switch to Netguard because Blokada doesn't allow completely block internet access for certain apps but Blokada is more user-friendly, so I came back, but I miss this option very much!

**Comment by Tobias-B-Besemer** [comment]
Just want to mention that if someone have a rooted device, he can use AFWall+ as a very good workaround...

**Comment by jwsp1** [comment]
Really miss this feature too! +1 for this 👍

**Comment by English106** [comment]
I would love to block internet access per app or at least while the screen is off.

**Comment by alexanderadam** [comment]
Just a hint for people who are using their phone with a Google Account:

If you didn't change it you can see what apps are currently tracked under [google.com/myactivity](https://myactivity.google.com/myactivity).
So when the feature is added you might get some inspiration what apps you might want to add.

I'm curious though why the responsible URL isn't tracked by the default lists.
@kar do you know why this might be?

**Comment by kar** [comment]
We don't manage the lists, you'd need to ask the maintainers.

**Comment by ovz93br43v7** [comment]
Is there any chance to get priority on this feature e.g. with a feature bounty donation?

**Comment by Ansem93** [comment]
I would love this feature too!
Since Netguard work the same way as blokada, shouldn't be possible to use their code to add this feature to blokada?

**Comment by OffifialMITX** [comment]
@Ansem93 I'm not a programmer, but, as far as I am aware, Netguard is very different in its codebase. It's pretty much as if you'd try to take a phone camera and replace its len

**Comment by OffifialMITX** [comment]
@ovz93br43v7 I second this, I would also be willing to 'invest' to help this feature manifest.

**Comment by ps100000** [comment]
This feature is being worked on but there are a **lot** of changes on critical pieces of code that have to be done so it might still take quite some time to implement this.

**Comment by ps100000** [comment]
**Bad news. Starting with Android 10 access to the info which app created a connection is blocked so until we find a new way to get this information its pointless to implement this feature as it would only support Android < 10.**

**Comment by OffifialMITX** [comment]
Oh… that's very unfortunate. Maybe it would be possible to utilize the work profile to isolate apps and create address profiles or something like that?

**Comment by animeking9901** [comment]
Ah I've seen apps use the work profile for isolation before,maybe that
would be a thing. IDK myself but perhaps it could be possible

**Comment by ovz93br43v7** [comment]
According to some Google Play comments this firewall app https://play.google.com/store/apps/details?id=app.greyshirts.firewall.beta should work with Android 10. 
This app work also via VPN but is not FOSS :-(
Probably the developer of this app can give you a hint how this is possible....

Am 25. Februar 2020 11:11:45 MEZ schrieb animeking9901 <notifications@github.com>:

**Comment by ps100000** [comment]
> Oh… that's very unfortunate. Maybe it would be possible to utilize the work profile to isolate apps and create address profiles or something like that?

Might have look at this option someday. For now I spend enough time trying to implement this feature just to realize it won't work this way. I'll come back to it. Just not sure when and how.

**Comment by OffifialMITX** [comment]
@ovz93br43v7 Do you (guys) think that decompilint that app would help? I've seen recently some news about a decompiler for Android apps being released, but ignored it because I don't know Java.

**Comment by ps100000** [comment]
> @ovz93br43v7 Do you (guys) think that decompilint that app would help? I've seen recently some news about a decompiler for Android apps being released, but ignored it because I don't know Java.

I'll have a look at it. Didn't do reversing Java for a bit so it should be a nice exercise. ^^

**Comment by jawz101** [comment]
How does NetGuard do it then?  Or AdGuard?

**Comment by OffifialMITX** [comment]
@jawz101 I'd suppose they are using the Android 9- way as well.

**Comment by OffifialMITX** [comment]
@jawz101 I'd suppose they are using the Android 9- way as well.

**Comment by jawz101** [comment]
I'm using Android 11 Developer Preview right now

**Comment by jeroen7s** [comment]
> Bad news. Starting with Android 10 access to the info which app created a connection is blocked

Wouldn't this also break the functionality to whitelist urls for a specific app?

**Comment by ps100000** [comment]
@OffifialMITX @jawz101 
I might have found the new way how other apps do it. So maybe this feature is still doable but it means even more work as it conflicts with Blokadas minSdkVersion. So I would have to implement the old and the new way.

**Comment by ps100000** [comment]
> Wouldn't this also break the functionality to whitelist urls for a specific app?

Whitelisting the hole app is still possible as this uses the normal VPN API.

**Comment by OffifialMITX** [comment]
@ps100000 I think they mean specific urls and not just whitelist apps in general?

**Comment by OffifialMITX** [comment]
@ps100000 
> means even more work

Again, is there some way we can support you? At least monetarily, or something. I, for one, don't have much, but would like to donate to a good cause.

**Comment by rugk** [comment]
To reiterate the scope of this issue here and to focus, here is what is said in the OP:
> I'd like to completely block internet access for certain apps.

So this is _not_ about:
* whitelisting (i.e. allowing all traffic)
* or about specific URLs (for an app)

This is just about **blocking all access** to the internet for one app.

IMHO, we should open other issues for other topics/feature requests.

**Comment by ps100000** [comment]
@OffifialMITX 
I'm not the main developer and I can only do this stuff in my freetime and I guess you can't donate me more freetime. ^^
But if you want to donate a few bugs to the main dev I'm sure he'll appreciate it. :)

**Comment by ps100000** [comment]
@rugk 
Sure but tbh it's all one pool of features connected to the same core functionality. If this part is finished the rest is relatively simple to do. :)

**Comment by jaypatelani** [comment]
Simple interface like this can be useful:

https://www.digitalcitizen.life/sites/default/files/gdrive/ios_block_internet/block_internet_4.png

**Comment by jaypatelani** [comment]
Simple interface like this can be useful:
![block_internet_4](https://user-images.githubusercontent.com/17174485/77842677-c43a5380-7184-11ea-8149-2787a9b787d9.png)

**Comment by AB1908** [comment]
Hi folks. Sorry to bump everyone but what's the status on this? From what I can gather from the history, Android 10 changes how this is handled so Blokada would need two features, one for the old OS and one for the new. Presumably, that's a major deterrent and this is on the backburner?

**Comment by erison15** [comment]
Just do it, please <3

**Comment by c933103** [comment]
Just like the app can already whitelist some apps and let all traffic from the app bypass blokada filter, it would be nice to be able to blacklist some app using a similar interface option and principle to block all internet access from those apps.

**Comment by abruenin** [comment]
Any news about this? Blokada has the best filtering. Other Firewalls like NetGuard can completely block apps, but their filtering is limited. Hopefully one of these will catch up where the other is better. I had no problem if this would become a paid feature and it does not have to be cheap. Hopefully someone will do it.

**Comment by alekksander** [comment]
if it solves anyone problem — tracker control (f-droid repo as well) has ability to deny internet access for chosen apps

**Comment by szaimen** [comment]
> if it solves anyone problem — tracker control (f-droid repo as well) has ability to deny internet access for chosen apps

Thanks but unfortunately you cannot use two apps that both use the android vpn service at the same time.

**Comment by ln-12** [comment]
Hey guys, I would also really want to see this feature as part of Blokada. Recently, I discovered the app BraveDNS which both provides custom DNS **and** a firewall. [It is open source](https://github.com/celzero/brave-android-app), so it should be easy to port this funtion to Blokada.

**Comment by AB1908** [comment]
> [It is open source](https://github.com/celzero/brave-android-app), so it should be easy to port this funtion to Blokada.

Please understand first off that I mean no hostility and am simply expressing my displeasure. It is not a personal attack in any shape or form and I apologise if it comes off as such.

That is a vastly oversimplified take of how FOSS development works. I say this as someone who has tried to dive into _simple_ codebases and failed. It is a known trope in the software dev world that "Why don't you do XYZ? It's so easy!" and it really grinds people's gears. It is understandable that a lot of people are unfamiliar with dev work but I must respectfully ask that you refrain from saying something similar in circles like these. Thanks for understanding!

**Comment by ln-12** [comment]
I am not unfamiliar with dev work at all. For me it is always a big relief to find some project containing code that I can reuse or at least use parts of. I took a quick look into BraveDNS and found some snippets which might be useful. [Here](https://github.com/celzero/brave-android-app/blob/a5eb607972c665842d46089202b3e38b9f1eadfe/app/src/main/java/com/celzero/bravedns/receiver/BraveScreenStateReceiver.kt#L60) they are collecting information about every app. The UID ([which is assigned at install time](https://source.android.com/security/app-sandbox)) is used as unique identifier. An internal variable is used to determine if an app should be blocked or not. [Here](https://github.com/celzero/brave-android-app/blob/a5eb607972c665842d46089202b3e38b9f1eadfe/app/src/main/java/com/celzero/bravedns/service/BraveVPNService.kt#L126) they check if an app should be blocked or  not. Finally, they find the UID of a connection by asking the [ConnectivityManager](https://developer.android.com/reference/android/net/ConnectivityManager#getConnectionOwnerUid(int,%20java.net.InetSocketAddress,%20java.net.InetSocketAddress)) (only API >=29, I know...).

While digging into Blokada, I found that a firewall feature seems to be on the TODO list as stated by [this](https://github.com/blokadaorg/blokada/blob/03c22a3ec953729967872c0ef40b20e1470e34a2/app/src/tun-local/kotlin/tunnel/Model.kt#L156) comment. Blokada also [already makes use](https://github.com/blokadaorg/blokada/search?q=ConnectivityManager) of the ConnectivityManager. The only thing I found not (by now) is where the firewall check should be implemented. But I think the developers of Blokada maybe can answer this question, so this feature could make it into the app in the near future, I would be really happy to see it. Maybe my comment can help.

**Comment by AB1908** [comment]
I apologise for making assumptions and I'm sure the devs would find your investigation useful. I have clearly misunderstood your intentions.

Indeed, it would be neat if this feature makes its way into the app. I would love to have it.

**Comment by ps100000** [comment]
Thanks for sharing your findings. In fact there are a few things happening in the background right now to get this feature into Blokada somewhere in the near future. :)

**Comment by ignoramous** [comment]
> I took a quick look into BraveDNS and found some snippets which might be useful. Here they are collecting information about every app. The UID (which is assigned at install time) is used as unique identifier.

Implementing a firewall the way RethinkDNS / BraveDNS does isn't ground breaking in anyway, the big issue is the UX (and follow-up customer support) around it, because it tends to breaks a lot of things, including wifi-calling and some video conferencing apps. That aside, most Chinese OEMs (which is like 60% of the Android phone market) like the Oppo companies (w/ Realme, OnePlus, Vivo), Xiaomi, the Lenovo companies (w/ Motorola) already come with a built-in, battery-efficient firewall.

RethinkDNS / BraveDNS doesn't implement a firewall for the sake of it. It is a security tool, first and foremost, and that warrants a "TCP/UDP network manager" of which firewall is _one_ part. Very different use-case to ad-blocking and/or privacy app that Blokada aims to be. Besides, [Google is allegedly implementing a firewall in AOSP](https://www.xda-developers.com/google-restricted-networking-mode-android-12/) themselves in Android 12.

**Comment by UstedLoco** [comment]
+1 on this feature request.

**Comment by furkun** [comment]
+1

**Comment by Ztumpie905509** [comment]
First of all, I am not an expert in this realm, I am just a normal user as everyone else expressing my opinion on this feature request. My arguments might not be as professional or accurate as of the above.

Only some phones have the ability to remove internet access for some apps via system settings as mentioned above (i.e. those Chinese OEMs).
For other devices that don't have this feature, some other VPN apps (such as [netguard](https://github.com/M66B/NetGuard)) are required to be installed.
This makes us cannot use blokada and those other apps simultaneously, which added some unnecessary steps when we have to switch between apps frequently.

However, if this firewall feature is added to the app, we are spared a lot of extra steps and provide a huge boost in user experience. We would definitely like to see this be implemented into the app in the near future.

**Comment by c933103** [comment]
The app currently already allow the vpn to apply to only certain apps.
Wouldn't it be possible to also only allow some of those VPNs to direct to an nonexistent tunnel at 127.0.0.1, or only return emptu response to those apps, thus effectively blocking them from accessing the internet?

**Comment by CrimsonFork** [comment]
I'm no developer, but I'm guessing that this would blokada unusable for the ad blocking, as to my knowledge, Android doesn't support having more than one VPN connection at a time.

**Comment by c933103** [comment]
> I'm no developer, but I'm guessing that this would blokada unusable for the ad blocking, as to my knowledge, Android doesn't support having more than one VPN connection at a time.
> 
> On November 4, 2021 1:07:53 AM UTC, c933103 ***@***.***> wrote:
> >The app currently already allow the vpn to apply to only certain apps.
> >Wouldn't it be possible to also only allow some of those VPNs to direct to an nonexistent tunnel at 127.0.0.1, or only return emptu response to those apps, thus effectively blocking them from accessing the internet?
> >
> >-- 
> >You are receiving this because you were mentioned.
> >Reply to this email directly or view it on GitHub:
> >https://github.com/blokadaorg/blokada/issues/49#issuecomment-960352374

I mean directing apps to different address for each indovidual app, within the blokada app
Like the whitelisting feature, after connected, the app can currently determine whether each apps are in the whitelist or not, and determine whether the app's request should be able to bypass the app's ad filtering. The same mechanism should be usable in returning empty response to request from some apps?

**Comment by tdbe** [comment]
> > I took a quick look into BraveDNS and found some snippets which might be useful. Here they are collecting information about every app. The UID (which is assigned at install time) is used as unique identifier.
> 
> Implementing a firewall the way RethinkDNS / BraveDNS does isn't ground breaking in anyway, the big issue is the UX (and follow-up customer support) around it, because it tends to breaks a lot of things, including wifi-calling and some video conferencing apps. That aside, most Chinese OEMs (which is like 60% of the Android phone market) like the Oppo companies (w/ Realme, OnePlus, Vivo), Xiaomi, the Lenovo companies (w/ Motorola) already come with a built-in, battery-efficient firewall.
> 
> RethinkDNS / BraveDNS doesn't implement a firewall for the sake of it. It is a security tool, first and foremost, and that warrants a "TCP/UDP network manager" of which firewall is _one_ part. Very different use-case to ad-blocking and/or privacy app that Blokada aims to be. Besides, [Google is allegedly implementing a firewall in AOSP](https://www.xda-developers.com/google-restricted-networking-mode-android-12/) themselves in Android 12.

Do not rely on google's "firewall". Google's firewall will route all traffic through them and then decide what to do with it after they've looked at it.

Of course it's a privacy feature to be able to block all internet traffic for an app! Think of it as a backup to the failure of Blokada's ability to figure out if you're being tracked or not. If I make a keyboard app that sends all keystrokes to my server, Blokada won't see that as a privacy breach. But we obviously would want to make sure our keyboard apps never ever talk to the internet.

See the point? We can't have 2 vpn apps at the same time, and without root, you're stuck with either a blokada VPN, or a firewall VPN. 90% of phones on the market do not have a builtin "firewall" or internet permission.

It's trivial to add a toggle per app to stop all traffic.

**Comment by tdbe** [comment]
I can't believe it's been 5 years and there are 0 solutions on the market that offer VPN + also firewall. This is the only reason I root my devices. But I can't root everyone's device. Please. Wtf. I could have done it myself in 5 years.... 🤔 I wonder if there's any forks with this feature actually

**Comment by alekksander** [comment]
@tdbe
not only it exist, but also is free for the time being. check out rethinkdns. not intuitive and bug-less, but does the job and so far it's the most configurable tool for non-rooted devices. ofc open sourced.

**Comment by alekksander** [comment]
> I can't believe it's been 5 years and there are 0 solutions on the market that offer VPN + also firewall. I would pay so much money for this... This is the only reason I root my devices. But I can't root everyone's device. Please. Wtf. I could have done it myself in 5 years.... thinking I wonder if there's any forks with this feature actually

not only it exist, but also is free for the time being. check out rethinkdns. not intuitive and bug-less, but does the job and so far it's the most configurable tool for non-rooted devices. ofc open sourced.

**Comment by tdbe** [comment]
> rethinkdns

This is not true. Rethink dns's firewall is a VPN. Which means I cannot use an actual VPN (e.g. ProtonVPN) and also a firewall at the same time. 

This is what I've been saying: nobody in the android space has merged a firewall and a vpn app in the same "vpn" solution. Not blokada, not rethink dns, not any vpn provider, nada.

I can't believe all these guys are "privacy experts" yet they cripple our privacy through the stubborn refusal to add a simple firewall switch

**Comment by Ansem93** [comment]
> I can't believe it's been 5 years and there are 0 solutions on the market that offer VPN + also firewall. I would pay so much money for this... This is the only reason I root my devices. But I can't root everyone's device. Please. Wtf. I could have done it myself in 5 years.... 🤔 I wonder if there's any forks with this feature actually

I switched to adGuard pro because it has both adblock and firewall. 30€ for a lifetime family licence (9 devices) is a good price

**Comment by tdbe** [comment]
> > I can't believe it's been 5 years and there are 0 solutions on the market that offer VPN + also firewall. I would pay so much money for this... This is the only reason I root my devices. But I can't root everyone's device. Please. Wtf. I could have done it myself in 5 years.... 🤔 I wonder if there's any forks with this feature actually
> 
> I switched to adGuard pro because it has both adblock and firewall. 30€ for a lifetime family licence (9 devices) is a good price

Interesting. Thanks for the heads-up. I hope the vpn is comparable to proton / pia / blokada. I'll look into it! :)

**Comment by maffe** [comment]
> I could have done it myself in 5 years

Maybe you could start working on it so we all can have it in 2026?

**Comment by alekksander** [comment]
> This is not true. Rethink dns's firewall is a VPN. Which means I cannot use an actual VPN (e.g. ProtonVPN, or Blokada Plus) and also a firewall at the same time.

i'm not sure, but in my setup it block applications from accessing internet completely. if i allow it in firewall it then filters with my lists.

**Comment by Ansem93** [comment]
> > > I can't believe it's been 5 years and there are 0 solutions on the market that offer VPN + also firewall. I would pay so much money for this... This is the only reason I root my devices. But I can't root everyone's device. Please. Wtf. I could have done it myself in 5 years.... 🤔 I wonder if there's any forks with this feature actually
> > 
> > 
> > I switched to adGuard pro because it has both adblock and firewall. 30€ for a lifetime family licence (9 devices) is a good price
> 
> Interesting. Thanks for the heads-up. I hope the vpn is comparable to proton / pia / blokada. I'll look into it! :)
> 
> There's no info on being audited, not open source, and they apparently log when and how much you're using the vpn. https://www.reddit.com/r/Adguard/comments/lkiwah/looking_for_a_decent_adguard_vpn_review/ Honestly still better than blokada without a fiewwall, and it does hugely pain me to say this..

I only use the app, not the vpn.
I would still using blokada but since it miss the firewall feature I had to switch :/

**Comment by arjan-s** [comment]
I switched to TrackerControl recently (completely free: https://f-droid.org/packages/net.kollnig.missioncontrol.fdroid). Not interesting for @tdbe because it doesn't have an actual VPN, but perhaps for others like me who want to combine ads/trackers blocking and firewall.

**Comment by AdKiller** [comment]
BLOCK is in the name... and yet it can not block an app from the internet?! that's disappointing...

**Comment by CrimsonFork** [comment]
Damn, second oldest still open issue, and the one with most comments by far. Brought to you all by yours truly.

**Comment by tdbe** [comment]
I always hate to say anything even remotely in the realm of conspiracy theory because usually there's simpler explanations.
But considering there are so many great VPN providers: blokada, mullvad, proton, nord, pia, etc. All of these have Split Tunnelling even.

And NONE OF THEM have a feature to just stop an app from the internet. And, none of them provide any technical reason/feedback as to why this might be problematic to implement. (it can't be difficult, because we already have per-app vpn firewalls)

And we'd gladly pay huge moneys for this feature. So wtf is google or some manufacturers breathing down their necks or something??? (android is supposed to provide per-app internet revocation permission, except it's removed from all official flavours, only cyanogenmod used to have it)

**Comment by c933103** [comment]
> I always hate to say anything even remotely in the realm of conspiracy because usually there's simpler explanations. But considering there are so many great VPN providers: blokada, mullvad, proton, nord, pia, etc. All of these have Split Tunnelling even.
> 
> And NONE OF THEM have a feature to just stop an app from the internet. And, none of them provide any technical reason/feedback as to why this might be problematic to implement. (it can't be difficult, because we already have per-app vpn firewalls)
> 
> And we'd gladly pay huge moneys for this feature. So wtf is google or some manufacturers breathing down their necks or something??? (android is supposed to provide per-app internet revocation permission, except it's removed from all official flavours, only cyanogenmod used to have it)

There are/were firewall apps that provide per-app blocking according to my understanding, but since they aren't VPN and they use VPN function of Android to loop all web traffic back to the app itself for filtering, those apps cannot be used together with actual VPN services. That indicate this is at least physically possible to block individual app access to internet, and that there aren't barriers that block such function from being implemented in an app.

**Comment by alekksander** [comment]
Actually invisible pro has an option to block apps from internet (they don't brag about it for some reason) and to work in VPN or PROXY mode. It also includes DNScrypt, I2P and TOR. It's quite batt. friendly as well. Because of the proxy mode its likely a golden gral youre looking for.

**Comment by maverick74** [comment]
> I switched to TrackerControl recently (completely free: https://f-droid.org/packages/net.kollnig.missioncontrol.fdroid). Not interesting for @tdbe because it doesn't have an actual VPN, but perhaps for others like me who want to combine ads/trackers blocking and firewall.

Yeah, I switched to TrackerControl as well and it's great. It works just as intended!!!

**Comment by DI555** [comment]
are any good news about firewall implementation???

**Comment by xgsx-8** [comment]
I would also like to know if there is any news about (DI555) firewall!?

**Comment by xgsx-8** [comment]
I would also like to know if there is any news about (DI555) blokada firewall!?
